src/execution/sql/sorter.cpp

```diff
@@ -0,0 +1,343 @@
+#include "execution/sql/sorter.h"
+
+#include <algorithm>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include <tbb/tbb.h>  // NOLINT
+
+#include "llvm/ADT/STLExtras.h"
+
+#include "ips4o/ips4o.hpp"
+
+#include "execution/sql/thread_state_container.h"
+#include "execution/util/stage_timer.h"
+#include "loggers/execution_logger.h"
+
+namespace terrier::sql {
+
+Sorter::Sorter(MemoryPool *memory, ComparisonFunction cmp_fn, u32 tuple_size)
+    : tuple_storage_(tuple_size, MemoryPoolAllocator<byte>(memory)),
+      owned_tuples_(memory),
+      cmp_fn_(cmp_fn),
+      tuples_(memory),
+      sorted_(false) {}
+
+Sorter::~Sorter() = default;
+
+byte *Sorter::AllocInputTuple() {
+  byte *ret = tuple_storage_.append();
+  tuples_.push_back(ret);
+  return ret;
+}
+
+byte *Sorter::AllocInputTupleTopK(UNUSED u64 top_k) { return AllocInputTuple(); }
+
+void Sorter::AllocInputTupleTopKFinish(const u64 top_k) {
+  // If the number of buffered tuples is less than top_k, we're done
+  if (tuples_.size() < top_k) {
+    return;
+  }
+
+  // If the number of buffered tuples matches tok_k, let's build the heap. Note,
+  // this will only ever be done once!
+  if (tuples_.size() == top_k) {
+    BuildHeap();
+    return;
+  }
+
+  //
+  // We've buffered one more tuple than should be in the top-K, so we may need
+  // to reorder the heap. Check if the most recently inserted tuple belongs in
+  // the heap.
+  //
+
+  const byte *last_insert = tuples_.back();
+  tuples_.pop_back();
+
+  const byte *heap_top = tuples_.front();
+
+  if (cmp_fn_(last_insert, heap_top) <= 0) {
+    // The last inserted tuples belongs in the top-k. Swap it with the current
+    // maximum and sift it down.
+    tuples_.front() = last_insert;
+    HeapSiftDown();
+  }
+}
+
+void Sorter::BuildHeap() {
+  const auto compare = [this](const byte *left, const byte *right) { return cmp_fn_(left, right) < 0; };
+  std::make_heap(tuples_.begin(), tuples_.end(), compare);
+}
+
+void Sorter::HeapSiftDown() {
+  const u64 size = tuples_.size();
+  u32 idx = 0;
+
+  const byte *top = tuples_[idx];
+
+  while (true) {
+    u32 child = (2 * idx) + 1;
+
+    if (child >= size) {
+      break;
+    }
+
+    if (child + 1 < size && cmp_fn_(tuples_[child], tuples_[child + 1]) < 0) {
+      child++;
+    }
+
+    if (cmp_fn_(top, tuples_[child]) >= 0) {
+      break;
+    }
+
+    std::swap(tuples_[idx], tuples_[child]);
+    idx = child;
+  }
+
+  tuples_[idx] = top;
+}
+
+void Sorter::Sort() {
+  // Exit if the input tuples have already been sorted
+  if (is_sorted()) {
+    return;
+  }
+
+  // Exit if there are no input tuples
+  if (tuples_.empty()) {
+    return;
+  }
+
+  // Time it
+  util::Timer<std::milli> timer;
+  timer.Start();
```

Comment:
@pmenon, should these timers be wrapped in conditionals so that they aren't built in release mode where their results aren't used?  If not, should we be capturing the results so that the metrics can be aggregated in the future?

Reply:
They'll most likely be optimized out in release. It's header-only all transitive uses are marked unused.